m3/reporter.go

```diff
@@ -109,16 +112,16 @@ type reporter struct {
 	calcProto       thrift.TProtocol
 	client          *m3thrift.M3Client
 	commonTags      []m3thrift.MetricTag
+	done            atomic.Bool
 	freeBytes       int32
 	metCh           chan sizedMetric
-	processors      sync.WaitGroup
+	now             atomic.Int64
+	overheadBytes   int32
+	pending         atomic.Uint64
 	resourcePool    *resourcePool
-	status          reporterStatus
-}
-
-type reporterStatus struct {
-	sync.RWMutex
-	closed bool
+	stringInterner  *cache.StringInterner
+	tagCache        *cache.TagCache
```

Comment:
Do we assume that tags will be the same during the lifetime of reporter so we don't need to expire/evict cached entries?

Reply:
tally so far had no LRU eviction or anything for scopes, so all that stayed in memory forever, which is bad for people who tag their metrics with UUIDs.

Reply:
Yes and no - reporters are scope-agnostic, so subscopes can continue to cause reporters to allocate, but if we assume that there is at least 2 metrics on every subscope, this pulls ahead pretty much immediately. Right now scopes hold onto the memory that cached reporters allocate indefinitely, which means that as soon as you allocate more than one metric on the same scope, you're re-allocating the same effective storage for every subsequent metric. So while linear growth is still an (eventual) problem, this at least substantially raises the ceiling (and we already would have hit limits sooner without this because we'd have been holding on to the extra storage).

There's another PR open (not planning to merge it yet) where it adds TTL-based metric eviction, which would feed into this to also evict cached tags if there were no more places that it was used.